db/view/view.cc

```diff
@@ -1719,37 +1722,150 @@ bool should_generate_view_updates_on_this_shard(const schema_ptr& base, const lo
 //
 // If the keyspace's replication strategy is a NetworkTopologyStrategy,
 // we pair only nodes in the same datacenter.
-// If one of the base replicas also happens to be a view replica, it is
-// paired with itself (with the other nodes paired by order in the list
+//
+// When use_legacy_self_pairing is enabled, if one of the base replicas
+// also happens to be a view replica, it is paired with itself
+// (with the other nodes paired by order in the list
 // after taking this node out).
 //
+// If the table uses tablets and the replication strategy is NetworkTopologyStrategy
+// and the replication factor in the node's datacenter is a multiple of the number
+// of racks in the datacenter, then pairing is rack-aware.  In this case,
+// all racks have the same number of replicas, and those are never migrated
+// outside their racks. Therefore, the base replicas are naturally paired with the
+// view replicas that are in the same rack, based on the ordinal position.
+// Note that typically, there is a single replica per rack and pairing is trivial.
+//
 // If the assumption that the given base token belongs to this replica
 // does not hold, we return an empty optional.
-static std::optional<gms::inet_address>
+std::optional<gms::inet_address>
 get_view_natural_endpoint(
+        locator::host_id me,
         const locator::effective_replication_map_ptr& base_erm,
         const locator::effective_replication_map_ptr& view_erm,
-        bool network_topology,
+        const locator::abstract_replication_strategy& replication_strategy,
         const dht::token& base_token,
         const dht::token& view_token,
         bool use_legacy_self_pairing,
+        bool use_tablets_rack_aware_view_pairing,
         replica::cf_stats& cf_stats) {
     auto& topology = base_erm->get_token_metadata_ptr()->get_topology();
-    auto me = topology.my_host_id();
-    auto my_datacenter = topology.get_datacenter();
+    auto& view_topology = view_erm->get_token_metadata_ptr()->get_topology();
+    auto& my_location = topology.get_location(me);
+    auto& my_datacenter = my_location.dc;
+    auto* network_topology = dynamic_cast<const locator::network_topology_strategy*>(&replication_strategy);
+    auto rack_aware_pairing = use_tablets_rack_aware_view_pairing && network_topology &&
+            (network_topology->get_replication_factor(my_datacenter) % topology.get_datacenter_racks().at(my_datacenter).size()) == 0;
```

Comment:
@bhalevy Is keeping this really a requirement? I have a feeling that determining co-location of replicas of sibling tablets will be really hard with this. I really don't understand yet why we cannot simply sort the base and view replica sets (by rack id) and rely on the order in the sorted list to determine pairing. If RF=3 and racks=2, it can happen base will have a higher load on rack 1, and view higher load on rack 0, yet there will be 1 pair that will be within a rack. When RF=3 and racks=3, then sorting will already provide the pairing within each rack, or at least eventually, until balancer moves each replica into a different rack. It is still rack aware, regardless.

I think we need to formalize the pairing algorithm and probably move it to locator (or somewhere else), such that we won't have to duplicate knowledge when taking it into account in the load balancer for merge.

Reply:
I'd really appreciate if you could explain to me why the above cannot work. If we can get rid of RF % rack check, load balancer can simply look at the rack aware flag to determine co-location.

Reply:
using stable sort by rack can work but it's sub-optimal since all racks between the shift in the first rack that has the imbalance until it's "corrected" by the mirror imbalance on the view, are misaligned.  But it's a theoretical concern that should rarely happen in the firld.  Maybe when renaming a rack via rolling restart or adding a new rack and decommissioning an existing one.

As for formaluzing it in locator, how would tablets merging use the pairing algorithm?

Reply:
@bhalevy I need some way for balancer to know whether it will have to sort replica sets by rack or not. So that it can determine whether replicas have same ordinal

Let me give you an example

assume n1=rack1, n2=rack2, n3=rack

tablet 0 {n1, n2, n3}
tablet 1 {n1, n3, n2}

even though n2 in tablet 0 and 1 have different index in the replica list, they still have same ordinal as used by rack-aware pairing. so they can be considered co-located and merged.

either balancer sorts itself the replica list based on the feature flag, or some function in locator does it for me also based on the flag. Maybe merge can assume rack aware pairing and it works easily when RF=#_of_racks.

Reply:
What about the single rack case?
Replicas would be in random node order (intentionally)

Reply:
> What about the single rack case?
> Replicas would be in random node order (intentionally)

But with stable sort we preserve the order no? Since all replicas share the same rack.

Reply:
The reason I keep pushing for this is for more consistent behavior.

If you have RF=3, and 2 racks, you still want replica sets sorted by rack id.

for example, let's assume

rack1 { N10, N11 }, rack 2 { N20, N21}

tablet 0 { N20, N10, N11 }
tablet 1 { N10, N20, N21 }

the above layout is possible, so N20 of tablet 0 is paired with N10 of tablet 1.

if you sort, then we have

tablet 0 { N10, N11, N20 }
tablet 1 { N10, N20, N21 }

later, if we add another rack, it's easier to maintain the pairing which will switch to rack aware pairing definitely.

then I think it would be good that we ditch simple rack aware pairing thing, to provide a seamless transition from rack2 to rack3, pairing is consistent across the change. Also makes my life easier in merge area :-)

Reply:
> The reason I keep pushing for this is for more consistent behavior.
> 
> If you have RF=3, and 2 racks, you still want replica sets sorted by rack id.
> 
> for example, let's assume
> 
> rack1 { N10, N11 }, rack 2 { N20, N21}
> 
> tablet 0 { N20, N10, N11 }
> tablet 1 { N10, N20, N21 }
> 
> the above layout is possible, so N20 of tablet 0 is paired with N10 of tablet 1.

But complex rack-aware pairing (in the latest version here) does sort by rack, so pairing would be:
N10 -> N10
N11 -> N20
N20 -> N21

> if you sort, then we have
> 
> tablet 0 { N10, N11, N20 }
> tablet 1 { N10, N20, N21 }
> 
> later, if we add another rack, it's easier to maintain the pairing which will switch to rack aware pairing definitely.
> 
> then I think it would be good that we ditch simple rack aware pairing thing, to provide a seamless transition from 2 racks to 3 racks, pairing is consistent across the change. Also makes my life easier in merge area :-)

I don't understand why the simple rack-aware pairing would cause inconsistencies if a rack is added.
Say we start with 2 racks and RF=2:
tablet 0: { N10, N20 }
tablet 1: { N11, N21 }

Simple rack-aware pairing would pair:
N10 -> N11
N20 -> N21

Then we add a 3rd rack, and load balancing modifies the map to:
tablet 0: { N30, N20 }
tablet 1: { N11, N21 }

The complex rack-aware pairing with sorting by rack, will result in:
N20 -> N11
N30 -> N21

Which is inconsistent with the simple rack-aware pairing

My previous best-match rack-aware pairing proposal would result in:
N20 -> N21
N30 -> N11

which extends simple rack-aware pairing.

Reply:
we can land in the replica sets below, if we start with RF=1, and later bump RF to 2.

tablet 0: { N20, N10 }
tablet 1: { N11, N21 }

RF=2, rack=2, so with simple rack aware pairing,

N20 -> N11
N10 -> N21

now if we bump RF to 3, we move to complex pairing

tablet 0: { N20, N10, ... }
tablet 1: { N11, N21, ... }

becomes (when sorted):

tablet 0: { N10, N20, ... }
tablet 1: { N11, N21, ... }

N10 -> N11
N20 -> N21

so the problem seems to come when switching from simple to complex, and in this setup, also the other way around.

are you reluctant to remove simple mode as an optimization due to concern of cost of sort for the common scenario?

Reply:
> we can land in the replica sets below, if we start with RF=1 (and 2 racks), and later bump RF to 2.
> 
> tablet 0: { N20, N10 } tablet 1: { N11, N21 }
> 
> RF=2, rack=2, so with simple rack aware pairing,
> 
> N20 -> N11
> N10 -> N21

No.  You either got me wrong, or I have a bug in my implementation.
Simple rack-aware pairing first filers both base and view replicas by rack,
and only withing the filtered endpoint it pairs in ordinal order, so pairing should actually be:

N20 -> N21
N10 -> N11

> are you reluctant to remove simple mode as an optimization due to concern of cost of sort for the common scenario?

Yeah.  I think the simple code is correct and more efficient than the complex mode that requires sorting by rack.
And since it is expected to be the prevalent case, especially in the cloud, we better keep it optimized.